flowredux/src/commonMain/kotlin/com/freeletics/flowredux/sideeffects/SideEffect.kt

```diff
@@ -63,15 +92,62 @@ internal class SideEffectBuilder<InputState : S, S, A>(
     fun build() = builder()
 }
 
-internal typealias GetState<S> = () -> S
+internal class ManagedSideEffect<InputState : S, S, A>(
+    private val builder: SideEffectBuilder<InputState, S, A>,
+    private val scope: CoroutineScope,
+    private val getState: GetState<S>,
+    private val stateChanges: SendChannel<ChangedState<S>>,
+) {
 
-internal fun <InputState : S, S, A> SideEffect<InputState, S, A>.produceStateGuarded(
-    actions: Flow<Action<A>>,
-    getState: GetState<S>,
-): Flow<ChangedState<S>> {
-    return produceState(actions, getState).map { guardWithIsInState(it, this) }
+    private var current: Current? = null
+
+    suspend fun sendStateChange(state: S) {
+        if (builder.isInState.check(state)) {
+            var current = current
+            if (current == null) {
+                val currentSideEffect = builder.build()
+                val currentJob = scope.launch {
+                    currentSideEffect.produceState(getState).collect {
+                        // the side effect is in state might be more specific than the one in the builder
+                        stateChanges.send(guardWithIsInState(it, currentSideEffect))
+                    }
+                }
+                current = Current(currentJob, currentSideEffect)
+                this.current = current
+            }
+
+            current.sideEffect.sendState(state)
+        }
+    }
+
+    suspend fun cancelIfNeeded(state: S) {
+        val current = current
+        if (current != null) {
+            if (!current.sideEffect.isInState.check(state)) {
+                current.job.cancel(StateChangeCancellationException())
```

Comment:
Having this exception here for the cancellation will mainly allow us to use it in tests to see if a block was properly cancelled.

Reply:
Not 100% sure if I understand how you will get the exception to be consumed inside the tests, but I guess I will see then in one of the next PRs.

Reply:
When we test the cancellation of a block we can now do

Reply:
When we test cancellation of a block we can now do 
```
var caughtThrowable: Throwable? = null

//...
on<Foo> {
  try {
    awaitCancellation()
    assertFail()
  }  catch (t: Throwable) {
  }
}

// ...
assertEquals(awaitItem(), TheStateThatCancelsTheBlockAbove)
assertEquals(caughtThrowable, StateChangeCancellationException())
```
So we're not just checking that it was cancelled but we're checking that it was cancelled by us.